### PR TITLE
Consolidation of Phase-2 e/g object interface

### PR DESCRIPTION
**Pull request by cerminar**
#### PR description:

This PR addresses several issues with the interface of `TkElectron` and `TkEm` objects and the HW objects used by the Phase2 GT emulator:

- add ID score to both the hardware and CMSSW electron objects;
- add interface to retrieve object in GT hw format;
- remove Ref. to `Egamma` standalone objects since they are not the actual ancestors due to architectural choice in L1T hardware. The ref is replaced by an `edm::Ptr` to the calorimeter primitive (different in barrel and endcap).
- further cleanup of the `TkElectron` interface and of data members;
- charge is now populated for all `TkElectrons`;
- change charge definition for GT electrons in accordance with new convention;
- use non-PV-based photon isolation in hardware objects; 
- eta and phi at vertex (from the track) are now used for the `TkElectron` coordinates.

*NOTE*: backward compatibility is maintained via IO rule in the reflex dictionary.

#### PR validation:

Successfully tested reading files produced with previous versions of the object data formats.

### CODE REVIEW

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc [hunk 1]
```diff
@@ -334,49 +314,40 @@ void L1TCtL2EgProducer::endJob() {
...
   emu.srcTrack = nullptr;
-  auto refEg = tkele.EGRef();
-  const auto newref = refRemapper.old2newRefMap.find(refEg);
-  if (newref != refRemapper.old2newRefMap.end()) {
-    refEg = newref->second;
-  }
-  refRemapper.origRefAndPtr.push_back(std::make_pair(refEg, tkele.trkPtr()));
-  emu.sta_idx = refRemapper.origRefAndPtr.size() - 1;
+
+  // FIXME: this is hugly
```
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc]: ?

File: DataFormats/L1TParticleFlow/interface/egamma.h [hunk 2]
```diff
@@ -87,15 +87,16 @@ namespace l1ct {
     z0_t hwZ0;
     tkdeta_t hwDEta;  // relative to the region center, at calo
     tkdphi_t hwDPhi;  // relative to the region center, at calo
+    id_score_t hwIDScore;
```
Comment (gpetruc) [inline DataFormats/L1TParticleFlow/interface/egamma.h:90]: BTW the two comments above on hwDEta and hwDPhi are misleading

**Review by gpetruc — COMMENTED** [review]
Looks reasonable to me

File: DataFormats/L1TCorrelator/src/TkElectron.cc [hunk 3]
```diff
@@ -11,14 +11,14 @@ using namespace l1t;
...
-
-{
+    : TkEm(p4, egCaloPtr, tkisol, -999), trkPtr_(trackPtr) {
   if (trkPtr_.isNonnull()) {
     setTrkzVtx(trkPtr()->POCA().z());
   }
 }
+
+TkElectron::TkElectron(const LorentzVector& p4, float tkisol)
+    : TkElectron(p4, edm::Ptr<L1Candidate>(nullptr, 0), edm::Ptr<L1TTTrackType>(nullptr, 0), tkisol) {}
```
Comment (gpetruc) [inline DataFormats/L1TCorrelator/src/TkElectron.cc]:
```suggestion
    : TkElectron(p4, edm::Ptr<L1Candidate>(), edm::Ptr<L1TTTrackType>(), tkisol) {}
```

File: DataFormats/L1TCorrelator/src/TkEm.cc [hunk 4]
```diff
@@ -5,17 +5,19 @@
...
+#include "DataFormats/Common/interface/RefToPtr.h"
 
 using namespace l1t;
 
 TkEm::TkEm() {}
 
-TkEm::TkEm(const LorentzVector& p4, const edm::Ref<EGammaBxCollection>& egRef, float tkisol)
-    : TkEm(p4, egRef, tkisol, -999) {}
+TkEm::TkEm(const LorentzVector& p4, float tkisol, float tkisolPV)
+    : TkEm(p4, edm::Ptr<L1Candidate>(nullptr, 0), tkisol, tkisolPV) {}
```
Comment (gpetruc) [inline DataFormats/L1TCorrelator/src/TkEm.cc]:
```suggestion
    : TkEm(p4, edm::Ptr<L1Candidate>(), tkisol, tkisolPV) {}
```

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc [hunk 5]
```diff
@@ -334,49 +314,40 @@ void L1TCtL2EgProducer::endJob() {
...
   emu.initFromBits(tkem.egBinaryWord<EGIsoObj::BITWIDTH>());
   emu.srcCluster = nullptr;
-  auto refEg = tkem.EGRef();
-  const auto newref = refRemapper.old2newRefMap.find(refEg);
-  if (newref != refRemapper.old2newRefMap.end()) {
-    refEg = newref->second;
-  }
-  refRemapper.origRefAndPtr.push_back(std::make_pair(refEg, edm::Ptr<RefRemapper::L1TTTrackType>(nullptr, 0)));
-  emu.sta_idx = refRemapper.origRefAndPtr.size() - 1;
+  constituentsPtrs.push_back(std::make_pair(tkem.egCaloPtr(), edm::Ptr<L1TTTrackType>(nullptr, 0)));
```
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc]:
```suggestion
  constituentsPtrs.push_back(std::make_pair(tkem.egCaloPtr(), edm::Ptr<L1TTTrackType>()));
```

**Review by gpetruc — CHANGES_REQUESTED** [review]
But perhaps better to use the standard defaultconstructor for null Ptrs rather than the constructor for a temporary ptr object with a null ptr

File: DataFormats/L1TParticleFlow/interface/egamma.h (hunk 2 above)
Comment (cerminar) [reply DataFormats/L1TParticleFlow/interface/egamma.h:90]: removed

**Comment by cerminar** [comment]
PR to master is https://github.com/cms-sw/cmssw/pull/43317

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no files with code format issues!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by epalencia** [comment]
Tagged as `phase2-l1t-1330pre3_v6`.

**Pull request merged**